rollup/l1/reader.go

```diff
@@ -0,0 +1,380 @@
+		l1CommitBatchEventSignature:   ScrollChainABI.Events[commitBatchEventName].ID,
+		l1RevertBatchEventSignature:   ScrollChainABI.Events[revertBatchEventName].ID,
+		l1FinalizeBatchEventSignature: ScrollChainABI.Events[finalizeBatchEventName].ID,
+	}
+
+	return &reader, nil
+}
+
+func (r *Reader) FinalizedL1MessageQueueIndex(blockNumber uint64) (uint64, error) {
+	data, err := r.l1MessageQueueABI.Pack(nextUnfinalizedQueueIndex)
+	if err != nil {
+		return 0, fmt.Errorf("failed to pack %s: %w", nextUnfinalizedQueueIndex, err)
+	}
+
+	result, err := r.client.CallContract(r.ctx, ethereum.CallMsg{
+		To:   &r.config.L1MessageQueueAddress,
+		Data: data,
+	}, new(big.Int).SetUint64(blockNumber))
+	if err != nil {
+		return 0, fmt.Errorf("failed to call %s: %w", nextUnfinalizedQueueIndex, err)
+	}
+
+	var parsedResult *big.Int
+	if err = r.l1MessageQueueABI.UnpackIntoInterface(&parsedResult, nextUnfinalizedQueueIndex, result); err != nil {
+		return 0, fmt.Errorf("failed to unpack result: %w", err)
+	}
+
+	next := parsedResult.Uint64()
+	if next == 0 {
+		return 0, nil
+	}
+
+	return next - 1, nil
+}
+
+func (r *Reader) LatestFinalizedBatch(blockNumber uint64) (uint64, error) {
+	data, err := r.scrollChainABI.Pack(lastFinalizedBatchIndex)
+	if err != nil {
+		return 0, fmt.Errorf("failed to pack %s: %w", lastFinalizedBatchIndex, err)
+	}
+
+	result, err := r.client.CallContract(r.ctx, ethereum.CallMsg{
+		To:   &r.config.ScrollChainAddress,
+		Data: data,
+	}, new(big.Int).SetUint64(blockNumber))
+	if err != nil {
+		return 0, fmt.Errorf("failed to call %s: %w", lastFinalizedBatchIndex, err)
+	}
+
+	var parsedResult *big.Int
+	if err = r.scrollChainABI.UnpackIntoInterface(&parsedResult, lastFinalizedBatchIndex, result); err != nil {
+		return 0, fmt.Errorf("failed to unpack result: %w", err)
+	}
+
+	return parsedResult.Uint64(), nil
+}
+
+// GetLatestFinalizedBlockNumber fetches the block number of the latest finalized block from the L1 chain.
+func (r *Reader) GetLatestFinalizedBlockNumber() (uint64, error) {
+	header, err := r.client.HeaderByNumber(r.ctx, big.NewInt(int64(rpc.FinalizedBlockNumber)))
+	if err != nil {
+		return 0, err
+	}
+	if !header.Number.IsInt64() {
+		return 0, fmt.Errorf("received unexpected block number in L1Client: %v", header.Number)
+	}
+	return header.Number.Uint64(), nil
+}
+
+// FetchBlockHeaderByNumber fetches the block header by number
+func (r *Reader) FetchBlockHeaderByNumber(blockNumber uint64) (*types.Header, error) {
+	return r.client.HeaderByNumber(r.ctx, big.NewInt(int64(blockNumber)))
+}
+
+// FetchTxData fetches tx data corresponding to given event log
+func (r *Reader) FetchTxData(txHash, blockHash common.Hash) ([]byte, error) {
+	tx, err := r.fetchTx(txHash, blockHash)
+	if err != nil {
+		return nil, err
+	}
+	return tx.Data(), nil
+}
+
+// FetchTxBlobHash fetches tx blob hash corresponding to given event log
+func (r *Reader) FetchTxBlobHash(txHash, blockHash common.Hash) (common.Hash, error) {
```

Comment:
Why only return one blob hash?

Reply:
Yeah I missed this. It is fixed in this follow up PR as we need to access all blob hashes: https://github.com/scroll-tech/go-ethereum/pull/1103/files#diff-69957add1b7f17ea35a71a72df600e3b4b17cc903320fe5ad603aa1eda1f351cR147